File: codeforces/1490G-old-floppy-drive/main.py
import math
T = int(input())
for t in range(T):
	N, M = map(int, input().split())
	cmax = list(map(int, input().split()))
	total = cmax[0]
	for i in range(1, N):
		total += cmax[i]
		cmax[i] = max(total, cmax[i-1])
	for x in map(int, input().split()):
		l, r = 0, 10**15
		while l < r:
			m = (l+r)//2
			srow = m//N*total
			if m < N:
				test = cmax[m] >= x
			elif total <= 0:
				test = cmax[N-1] >= x
			else:
				test = max(srow - total + cmax[N-1], srow + cmax[m%N]) >= x
			if test:
				r = m
			else:
				l = m+1
		print(l if l!=10**15 else -1, end=' ')
	print()

File: codeforces/1490G-old-floppy-drive/main.cpp
#include <cstdio>
#include <vector>
typedef long long LL;
void solve() {
	LL N, M;
	std::scanf("%lld %lld", &N, &M);
	std::vector<LL> cmax(1);
	std::scanf("%lld", &cmax[0]);
	LL total = cmax[0];
	for (LL i = 1, x; i < N; ++i) {
		std::scanf("%lld", &x);
		total += x;
		cmax.push_back(std::max(cmax.back(), total));
	}
	for (LL i = 0, x; i < M; ++i) {
		std::scanf("%lld", &x);
		LL l = 0, r = 1e9/total+5;
		LL row = 0;
		while (total > 0 && l < r) {
			LL m = l+(r-l)/2;
			if (m*total + cmax[N-1] >= x)
				r = m;
			else
				l = m+1;
			row = l;
		}
		l = 0, r = N, x -= row*total;
		while (l < r) {
			LL m = l+(r-l)/2;
			if (cmax[m] >= x)
				r = m;
			else
				l = m+1;
		}
		LL col = l;
		std::printf("%lld ", (col == N ? -1 : row*N + col));
	}
	std::printf("\n");
}
int main() {
	LL T;
	std::scanf("%lld", &T);
	for (LL t = 0; t < T; ++t)
		solve();
}
